external-crates/move/crates/move-compiler/src/cfgir/optimize/forwarding_jumps.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//**************************************************************************************************
+// OPTIMIZE FORWARDING JUMPS
+// -----------------------------
+// This optimization forwards and removes "forwarding" jumps within a function. For example, the
+// following code will jump to 1 just to immediately jump to 3:
+//
+// Label 0:
+//     JumpIf(<exp>, 1, 2)
+// Label 1
+//     Jump 3
+// Label 2:
+//     ...
+//     Ret
+// Label 3:
+//     ...
+//
+// This optimization instead fowards these jumps, eliminating the other blocks:
+// Label 0:
+//     JumpIf(<exp>, 2, 1)
+// Label 1:
+//     ...
+//     Ret
+// Label 2:
+//     ...
+
+use crate::{
+    cfgir::{
+        ast::remap_labels,
+        cfg::{MutForwardCFG, CFG},
+    },
+    hlir::ast::{BasicBlocks, Command, Command_, FunctionSignature, Label, SingleType, Value, Var},
+    parser::ast::ConstantName,
+    shared::unique_map::UniqueMap,
+};
+
+use std::collections::{BTreeMap, BTreeSet};
+
+/// returns true if anything changed
+pub fn optimize(
+    _signature: &FunctionSignature,
+    _locals: &UniqueMap<Var, SingleType>,
+    _constants: &UniqueMap<ConstantName, Value>,
+    cfg: &mut MutForwardCFG,
+) -> bool {
+    let changed = optimize_(cfg.blocks_mut());
+    if changed {
+        cfg.recompute();
+        remap_in_order(cfg.start_block(), cfg.blocks_mut());
+    }
+    changed
+}
+
+fn optimize_(blocks: &mut BasicBlocks) -> bool {
+    let final_jumps = find_forwarding_jump_destinations(blocks);
+    optimize_forwarding_jumps(blocks, final_jumps)
+}
+
+type LabelMap = BTreeMap<Label, Label>;
+
+fn find_forwarding_jump_destinations(blocks: &BasicBlocks) -> LabelMap {
+    use Command_ as C;
+    let mut forwarding_jumps = BTreeMap::new();
+    for (label, block) in blocks.iter().filter(|(_, block)| block.len() == 1) {
+        if let Some(sp!(_, C::Jump { target, .. })) = block.iter().last() {
+            forwarding_jumps.insert(*label, *target);
+        }
+    }
+
+    // Computes the label map of forwarding jumps to their final destinations, collapsing any
+    // forwarding blocks between the starting label and final label. Note that we have to
+    // take care to detect cycles (fairly common in empty while loops). We also always look up the
+    // new label up in the final map to allow us to short-circuit work we've already done.
+    let mut final_jumps: LabelMap = BTreeMap::new();
+
+    for start in forwarding_jumps.keys() {
+        let mut target = *start;
+        let mut seen = BTreeSet::new();
+        while let Some(next_target) = forwarding_jumps.get(&target) {
```

Comment:
nit: Could you remember the path you went down, and then fix up the final jumps for all the `next_targets` you visit (i.e. path compression)?  That avoids this code being sensitive to iteration order.

Reply:
Sure, we actually already keep the seen list. We can just iterate it and make them all point to the final target.